pkg/wasm/cache_test.go

```diff
@@ -0,0 +1,234 @@
+			wantServerReqNum:     0,
+		},
+		{
+
+			name:                 "wrong checksum",
+			initialCachedModules: map[cacheKey]cacheEntry{},
+			fetchURL:             ts.URL,
+			purgeInterval:        DefaultWasmModulePurgeInteval,
+			wasmModuleExpiry:     DefaultWasmModuleExpiry,
+			checksum:             sha256.Sum256([]byte("wrongchecksum\n")),
+			wantErrorMsgPrefix:   fmt.Sprintf("module downloaded from %v has checksum %x, which does not match", ts.URL, dataCheckSum),
+			wantServerReqNum:     1,
+		},
+		{
+			// this might be common error in user configuration, that url was updated, but not checksum.
+			// Test that downloading still proceeds and error returns.
+			name: "different url same checksum",
+			initialCachedModules: map[cacheKey]cacheEntry{
+				{downloadURL: ts.URL, checksum: fmt.Sprintf("%x", dataCheckSum)}: {modulePath: fmt.Sprintf("%x.wasm", dataCheckSum)},
+			},
+			fetchURL:           ts.URL + "/different-url",
+			purgeInterval:      DefaultWasmModulePurgeInteval,
+			wasmModuleExpiry:   DefaultWasmModuleExpiry,
+			checksum:           dataCheckSum,
+			wantErrorMsgPrefix: fmt.Sprintf("module downloaded from %v/different-url has checksum", ts.URL),
+			wantServerReqNum:   1,
+		},
+		{
+			name: "purge on expiry",
+			initialCachedModules: map[cacheKey]cacheEntry{
+				{downloadURL: ts.URL, checksum: fmt.Sprintf("%x", dataCheckSum)}: {modulePath: fmt.Sprintf("%x.wasm", dataCheckSum)},
+			},
+			fetchURL:         ts.URL,
+			purgeInterval:    1 * time.Millisecond,
+			wasmModuleExpiry: 1 * time.Millisecond,
+			checksum:         dataCheckSum,
+			wantFileName:     fmt.Sprintf("%x.wasm", dataCheckSum),
+			wantServerReqNum: 1,
+		},
+		{
+			name: "cache hit before purge",
+			initialCachedModules: map[cacheKey]cacheEntry{
+				{downloadURL: ts.URL, checksum: fmt.Sprintf("%x", dataCheckSum)}: {modulePath: fmt.Sprintf("%x.wasm", dataCheckSum)},
+			},
+			fetchURL:         ts.URL,
+			purgeInterval:    50 * time.Millisecond,
+			wasmModuleExpiry: 50 * time.Millisecond,
+			checksum:         dataCheckSum,
+			wantFileName:     fmt.Sprintf("%x.wasm", dataCheckSum),
+			wantServerReqNum: 0,
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			tmpDir := t.TempDir()
+			cache := NewLocalFileCache(tmpDir, c.purgeInterval, c.wasmModuleExpiry)
+			defer close(cache.stopChan)
+			tsNumRequest = 0
+
+			cache.mux.Lock()
+			for k, m := range c.initialCachedModules {
+				filePath := filepath.Join(tmpDir, m.modulePath)
+				err := ioutil.WriteFile(filePath, []byte("data/\n"), 0644)
+				if err != nil {
+					t.Fatalf("failed to write initial wasm module file %v", err)
+				}
+				cache.modules[cacheKey{downloadURL: k.downloadURL, checksum: k.checksum}] =
+					cacheEntry{modulePath: filePath, last: time.Now()}
+			}
+			cache.mux.Unlock()
+
+			// Sleep 10 ms for purge on expiry testing
+			time.Sleep(10 * time.Millisecond)
```

Comment:
Generally sleep is not preferred, in prow we get absurd levels of contention and have tests regularly freeze for much longer than 10ms. Can we poll for success?

Reply:
Done.